File: src/main/java/com/nibado/projects/advent/y2015/Day04.java
package com.nibado.projects.advent.y2015;

import java.security.MessageDigest;

import static com.nibado.projects.advent.Util.printAnswer;

public class Day04 implements Runnable {
    private MessageDigest md5;
    private byte[] input;

    public void setInput(String input) {
        try {
            this.input = input.getBytes("UTF-8");
            md5 = MessageDigest.getInstance("md5");
        }
        catch(Exception e) {
            throw new RuntimeException(e);
        }
    }

    public static void main(String... argv) throws Exception {
        new Day04().run();
    }

    public int find(int zeroes) {
        for(int i = 0;i < Integer.MAX_VALUE;i++) {
            byte[] md5 = digest(i);

            for(int j = 0;j < zeroes;j++) {
                int mask = j % 2 == 0 ? 0xF0 : 0x0F;
                if((md5[j / 2] & mask) != 0) {
                    break;
                }

                if(j == zeroes - 1) {
                    return i;
                }
            }
        }

        return -1;
    }

    public byte[] digest(int i) {
        md5.update(input);
        return md5.digest(Integer.toString(i).getBytes());
    }

    @Override
    public void run() {
        setInput("iwrupvqb");
        int result = find(5);
        printAnswer(result);

        result =  find(6);
        printAnswer(result);
    }
}

File: src/main/kotlin/com/nibado/projects/advent/y2015/Day04.kt
package com.nibado.projects.advent.y2015

import com.nibado.projects.advent.*
import java.security.MessageDigest

object Day04 : Day {
    private val input = resourceString(2015, 4).trim().toByteArray()

    private fun find(zeroes: Int) =
        md5().first { (a, _) ->
            var ok = true
            for(i in 0 until zeroes) {
                val mask = if (i % 2 == 0) 0xF0 else 0x0F
                if(a[i / 2].toInt() and mask != 0) {
                    ok = false
                    break
                }
            }
            ok
        }

    private fun md5() = sequence<Pair<ByteArray, Int>> {
        val md5 = MessageDigest.getInstance("md5")

        for(i in 0 .. Int.MAX_VALUE) {
            md5.update(input)
            yield(md5.digest(i.toString().toByteArray()) to i)
        }
    }

    override fun part1() = find(5)
    override fun part2() = find(6)
}
